kernel/kernel-api/src/main/java/io/delta/kernel/internal/commitrange/CommitRangeFactory.java

```diff
@@ -139,7 +147,151 @@ private void logResolvedVersions(long startVersion, Optional<Long> endVersionOpt
+    // Note: currently this is all we allow in CommitRangeBuilder anyway, but in the future that
+    // could change
+    return ctx.logDatas.stream()
+        .filter(logData -> logData instanceof ParsedCatalogCommitData)
+        .map(logData -> (ParsedCatalogCommitData) logData)
+        .filter(deltaData -> deltaData.isFile())
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * Lists the _delta_log and combines the found published deltas with the catalog ratified deltas
+   * to compile a single contiguous list of deltas. Catalog commits take priority over published
+   * deltas when both are present for the same commit version. Returned deltas are guaranteed to
+   * start with startVersion, end with endVersion if endVersionOpt is non-empty, and are contiguous.
+   * Throws an exception if no deltas are found in the version range, or if startVersion or
+   * endVersion cannot be found.
+   */
+  private List<ParsedDeltaData> getDeltasForVersionRangeWithCatalogPriority(
+      Engine engine,
+      long startVersion,
+      Optional<Long> endVersionOpt,
+      List<ParsedCatalogCommitData> ratifiedCommits) {
+    // Get published deltas between startVersion and endVersionOpt
+    List<ParsedDeltaData> publishedDeltas =
+        getPublishedDeltasInVersionRange(engine, startVersion, endVersionOpt);
+
+    // Get ratified deltas between startVersion and endVersionOpt
+    List<ParsedDeltaData> ratifiedDeltas =
+        ratifiedCommits.stream()
+            .filter(
+                x ->
+                    x.getVersion() >= startVersion
+                        && x.getVersion() <= endVersionOpt.orElse(Long.MAX_VALUE))
+            .collect(Collectors.toList());
+
+    // Validate they are contiguous and valid (i.e. backfill is ordered)
+    validatePublishedPlusRatifiedDeltas(publishedDeltas, ratifiedDeltas);
+    List<ParsedDeltaData> combinedDeltas =
+        LogDataUtils.combinePublishedAndRatifiedDeltasWithCatalogPriority(
+            publishedDeltas, ratifiedDeltas);
+    validateVersionRange(combinedDeltas, startVersion, endVersionOpt);
+    return combinedDeltas;
+  }
+
+  /**
+   * Returns any published deltas found on the file-system within the version range provided and
+   * validates that they are contiguous. Returns a sorted and contiguous list of deltas for the
+   * version range, but does no validation that the list fills the range.
+   */
+  private List<ParsedDeltaData> getPublishedDeltasInVersionRange(
+      Engine engine, long startVersion, Optional<Long> endVersionOpt) {
+    final List<FileStatus> commitFiles =
+        listDeltaLogFilesAsIter(
+                engine,
+                Collections.singleton(FileNames.DeltaLogFileType.COMMIT),
+                tablePath,
+                startVersion,
+                endVersionOpt,
+                false /* mustBeRecreatable */)
+            .toInMemoryList();
+    List<ParsedDeltaData> publishedDeltas =
```

Comment:
can this be `ParsedPublishedDeltaData`?

Reply:
I think I can do this if I change the signature of `LogDataUtils.combinePublishedAndRatifiedDeltasWithCatalogPriority` and the other usage of it. Honestly, it makes sense for that fx to take in the specific sub-classes so I'm happy to do this.

Reply:
<img width="904" height="259" alt="Screenshot 2025-10-07 at 2 07 09 PM" src="https://github.com/user-attachments/assets/1b7eefea-a562-4ae6-b09a-a66a36239359" />

I think this change makes sense, but I will have to explicitly cast at least here. Do you think it's worth it to make this more `~compilish~`-correct?